ffi/src/transaction/transaction_id.rs

```diff
@@ -0,0 +1,182 @@
+use crate::error::ExternResult;
+use crate::handle::Handle;
+use crate::transaction::ExclusiveTransaction;
+use crate::{
+    ExternEngine, IntoExternResult, KernelStringSlice, OptionalValue, SharedExternEngine,
+    SharedSnapshot, TryFromStringSlice,
+};
+use delta_kernel::transaction::Transaction;
+use delta_kernel::{DeltaResult, Snapshot};
+use std::sync::Arc;
+
+/// Associates an app_id and version with a transaction. These will be applied to the table on commit.
+///
+/// # Returns
+/// A new handle to the transaction that will set the `app_id` version to `version` on commit
+///
+/// # Safety
+/// Caller is responsible for passing [valid][Handle#Validity] handles. CONSUMES TRANSACTION
+#[no_mangle]
+pub unsafe extern "C" fn with_transaction_id(
+    txn: Handle<ExclusiveTransaction>,
+    app_id: KernelStringSlice,
+    version: i64,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<Handle<ExclusiveTransaction>> {
+    let txn = unsafe { txn.into_inner() };
+    let engine = unsafe { engine.as_ref() };
+    with_transaction_id_impl(*txn, app_id, version).into_extern_result(&engine)
+}
+
+fn with_transaction_id_impl(
+    txn: Transaction,
+    app_id: KernelStringSlice,
+    version: i64,
+) -> DeltaResult<Handle<ExclusiveTransaction>> {
+    let app_id_string: DeltaResult<String> = unsafe { TryFromStringSlice::try_from_slice(&app_id) };
+    Ok(Box::new(txn.with_transaction_id(app_id_string?, version)).into())
```

Comment:
one thing i'm noticing with these builder FFIs - we are un-boxing (above) and re-boxing (here) every builder method that consumes self. maybe not worth optimizing _yet_ but seems non-ideal

Reply:
This sounds like it would be a follow up issue to this PR, right? Should we open an issue for it?